File: scripts/sync_contracts.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Canonical shared fragments are copied into each installed Skill so no runtime
# dependency on the repository root is introduced. Edit contracts/, then sync.
ROOT = File.expand_path("..", __dir__)
LOCAL_SKILLS = %w[focal compass flywheel soul].freeze
FRAGMENTS = %w[anchors severity evidence].freeze
abort "Usage: ruby scripts/sync_contracts.rb [--check]" unless (ARGV - ["--check"]).empty?
check = ARGV.include?("--check")
changes = {}
errors = []

targets = LOCAL_SKILLS.map { |skill| ["#{skill}/reference/review.md", FRAGMENTS] }
targets << ["product-judgement/SKILL.md", ["evidence"]]
targets.each do |path, fragments|
  content = File.read(File.join(ROOT, path), encoding: "UTF-8")
  original = content.dup
  fragments.each do |name|
    start = "<!-- BEGIN SHARED: #{name} -->"
    finish = "<!-- END SHARED: #{name} -->"
    pattern = /#{Regexp.escape(start)}\n.*?\n#{Regexp.escape(finish)}/m
    matches = content.scan(pattern)
    if matches.length != 1
      errors << "#{path}: expected exactly one #{name} fragment, found #{matches.length}"
      next
    end
    canonical = File.read(File.join(ROOT, "contracts", "#{name}.md"), encoding: "UTF-8").strip
    content = content.sub(pattern) { "#{start}\n#{canonical}\n#{finish}" }
  end
  changes[path] = content if content != original
end
abort errors.join("\n") unless errors.empty?
if check
  abort "Shared contracts are stale: #{changes.keys.join(', ')}. Run ruby scripts/sync_contracts.rb" unless changes.empty?
  puts "Shared evidence, score anchors, and severity contracts are current."
else
  changes.each { |path, content| File.write(File.join(ROOT, path), content) }
  puts "Synchronized #{changes.length} Skill file(s) from contracts/."
end

File: scripts/sync_contracts.py
#!/usr/bin/env python3
"""Synchronize canonical shared contract fragments into installed Skills."""

from __future__ import annotations

import re
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parent.parent
LOCAL_SKILLS = ["focal", "compass", "flywheel", "soul"]
FRAGMENTS = ["anchors", "severity", "evidence"]


def main(argv: list[str]) -> int:
    if any(argument != "--check" for argument in argv):
        print("Usage: uv run scripts/sync_contracts.py [--check]", file=sys.stderr)
        return 1
    check = "--check" in argv

    changes: dict[str, str] = {}
    errors: list[str] = []

    targets: list[tuple[str, list[str]]] = [
        (f"{skill}/reference/review.md", FRAGMENTS) for skill in LOCAL_SKILLS
    ]
    targets.append(("product-judgement/SKILL.md", ["evidence"]))

    for relative_path, fragments in targets:
        path = ROOT / relative_path
        content = path.read_bytes().decode("utf-8")
        original = content
        for name in fragments:
            start = f"<!-- BEGIN SHARED: {name} -->"
            finish = f"<!-- END SHARED: {name} -->"
            pattern = re.compile(re.escape(start) + r"\n.*?\n" + re.escape(finish), re.DOTALL)
            matches = pattern.findall(content)
            if len(matches) != 1:
                errors.append(f"{relative_path}: expected exactly one {name} fragment, found {len(matches)}")
                continue
            canonical = (ROOT / "contracts" / f"{name}.md").read_bytes().decode("utf-8").strip()
            content = pattern.sub(
                lambda _match: f"{start}\n{canonical}\n{finish}",
                content,
                count=1,
            )
        if content != original:
            changes[relative_path] = content

    if errors:
        print("\n".join(errors), file=sys.stderr)
        return 1

    if check:
        if changes:
            changed = ", ".join(changes)
            print(
                f"Shared contracts are stale: {changed}. Run uv run scripts/sync_contracts.py",
                file=sys.stderr,
            )
            return 1
        print("Shared evidence, score anchors, and severity contracts are current.")
        return 0

    for relative_path, content in changes.items():
        (ROOT / relative_path).write_bytes(content.encode("utf-8"))
    print(f"Synchronized {len(changes)} Skill file(s) from contracts/.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
